File: internal/core/events.go
// Package core provides loop event types for the loop engine.

package core

import (
	"fmt"
	"strings"
	"time"
)

// LoopStartEvent indicates the loop has started.
type LoopStartEvent struct {
	// Config is the loop configuration.
	Config *LoopConfig
}

// NewLoopStartEvent creates a new LoopStartEvent.
func NewLoopStartEvent(config *LoopConfig) *LoopStartEvent {
	return &LoopStartEvent{
		Config: config,
	}
}

// LoopCompleteEvent indicates the loop completed successfully.
type LoopCompleteEvent struct {
	// Result contains the loop result.
	Result *LoopResult
}

// NewLoopCompleteEvent creates a new LoopCompleteEvent.
func NewLoopCompleteEvent(result *LoopResult) *LoopCompleteEvent {
	return &LoopCompleteEvent{
		Result: result,
	}
}

// LoopFailedEvent indicates the loop failed.
type LoopFailedEvent struct {
	// Error is the error that caused the failure.
	Error error
	// Result contains partial loop result.
	Result *LoopResult
}

// NewLoopFailedEvent creates a new LoopFailedEvent.
func NewLoopFailedEvent(err error, result *LoopResult) *LoopFailedEvent {
	return &LoopFailedEvent{
		Error:  err,
		Result: result,
	}
}

// LoopCancelledEvent indicates the loop was cancelled by the user.
type LoopCancelledEvent struct {
	// Result contains partial loop result.
	Result *LoopResult
}

// NewLoopCancelledEvent creates a new LoopCancelledEvent.
func NewLoopCancelledEvent(result *LoopResult) *LoopCancelledEvent {
	return &LoopCancelledEvent{
		Result: result,
	}
}

// IterationStartEvent indicates an iteration has started.
type IterationStartEvent struct {
	// Iteration is the iteration number (1-based).
	Iteration int
	// MaxIterations is the maximum number of iterations.
	MaxIterations int
}

// NewIterationStartEvent creates a new IterationStartEvent.
func NewIterationStartEvent(iteration, maxIterations int) *IterationStartEvent {
	return &IterationStartEvent{
		Iteration:     iteration,
		MaxIterations: maxIterations,
	}
}

// IterationCompleteEvent indicates an iteration completed.
type IterationCompleteEvent struct {
	// Iteration is the iteration number (1-based).
	Iteration int
	// Duration is how long the iteration took.
	Duration time.Duration
}

// NewIterationCompleteEvent creates a new IterationCompleteEvent.
func NewIterationCompleteEvent(iteration int, duration time.Duration) *IterationCompleteEvent {
	return &IterationCompleteEvent{
		Iteration: iteration,
		Duration:  duration,
	}
}

// AIResponseEvent indicates AI response text was received.
type AIResponseEvent struct {
	// Text is the AI response text.
	Text string
	// Iteration is the current iteration number.
	Iteration int
}

// NewAIResponseEvent creates a new AIResponseEvent.
func NewAIResponseEvent(text string, iteration int) *AIResponseEvent {
	return &AIResponseEvent{
		Text:      text,
		Iteration: iteration,
	}
}

type ToolEvent struct {
	// ToolName is the name of the tool executed.
	ToolName string
	// Parameters are the tool parameters.
	Parameters map[string]interface{}
	// Iteration is the current iteration number.
	Iteration int
}

// Info returns a formatted string describing the tool execution based on parameters.
// This provides human-readable information about what the tool is doing.
func (e *ToolEvent) Info(emoji string) string {
	if len(e.Parameters) == 0 {
		return fmt.Sprintf("%s %s", emoji, e.ToolName)
	}

	var values []string
	for _, v := range e.Parameters {
		values = append(values, fmt.Sprintf("%v", v))
	}

	return fmt.Sprintf("%s %s: %s", emoji, e.ToolName, strings.Join(values, ", "))
}

// ToolExecutionEvent indicates a tool was executed.
type ToolExecutionEvent struct {
	ToolEvent

	// Result is the tool execution result.
	Result string
	// Error is any error that occurred during execution.
	Error error
	// Duration is how long the tool took to execute.
	Duration time.Duration
}

// NewToolExecutionEvent creates a new ToolExecutionEvent.
func NewToolExecutionEvent(toolName string, params map[string]interface{}, result string, err error, duration time.Duration, iteration int) *ToolExecutionEvent {
	return &ToolExecutionEvent{
		ToolEvent: ToolEvent{
			ToolName:   toolName,
			Parameters: params,
			Iteration:  iteration,
		},
		Result:   result,
		Error:    err,
		Duration: duration,
	}
}

// ToolExecutionStartEvent indicates a tool execution has started.
type ToolExecutionStartEvent struct {
	ToolEvent
}

// NewToolExecutionStartEvent creates a new ToolExecutionStartEvent.
func NewToolExecutionStartEvent(toolName string, params map[string]interface{}, iteration int) *ToolExecutionStartEvent {
	return &ToolExecutionStartEvent{
		ToolEvent: ToolEvent{
			ToolName:   toolName,
			Parameters: params,
			Iteration:  iteration,
		},
	}
}

// PromiseDetectedEvent indicates the promise phrase was found.
type PromiseDetectedEvent struct {
	// Phrase is the promise phrase that was detected.
	Phrase string
	// Source is where the promise was found (e.g., "ai_response", "tool_output").
	Source string
	// Iteration is the iteration number where promise was found.
	Iteration int
}

// NewPromiseDetectedEvent creates a new PromiseDetectedEvent.
func NewPromiseDetectedEvent(phrase, source string, iteration int) *PromiseDetectedEvent {
	return &PromiseDetectedEvent{
		Phrase:    phrase,
		Source:    source,
		Iteration: iteration,
	}
}

// ErrorEvent indicates an error occurred.
type ErrorEvent struct {
	// Error is the error that occurred.
	Error error
	// Iteration is the current iteration number (0 if not in iteration).
	Iteration int
	// Recoverable indicates if the error is recoverable.
	Recoverable bool
}

// NewErrorEvent creates a new ErrorEvent.
func NewErrorEvent(err error, iteration int, recoverable bool) *ErrorEvent {
	return &ErrorEvent{
		Error:       err,
		Iteration:   iteration,
		Recoverable: recoverable,
	}
}

File: src/Ralph.Cli/Core/Events.cs
// Loop event types for the loop engine.

namespace Ralph.Cli.Core;

/// <summary>
/// Base interface for all loop events.
/// </summary>
public interface ILoopEvent { }

/// <summary>
/// Indicates the loop has started.
/// </summary>
public sealed record LoopStartEvent(LoopConfig Config) : ILoopEvent;

/// <summary>
/// Indicates the loop completed successfully.
/// </summary>
public sealed record LoopCompleteEvent(LoopResult Result) : ILoopEvent;

/// <summary>
/// Indicates the loop failed.
/// </summary>
public sealed record LoopFailedEvent(Exception Error, LoopResult Result) : ILoopEvent;

/// <summary>
/// Indicates the loop was cancelled by the user.
/// </summary>
public sealed record LoopCancelledEvent(LoopResult Result) : ILoopEvent;

/// <summary>
/// Indicates an iteration has started.
/// </summary>
public sealed record IterationStartEvent(int Iteration, int MaxIterations) : ILoopEvent;

/// <summary>
/// Indicates an iteration completed.
/// </summary>
public sealed record IterationCompleteEvent(int Iteration, TimeSpan Duration) : ILoopEvent;

/// <summary>
/// Indicates AI response text was received.
/// </summary>
public sealed record AIResponseEvent(string Text, int Iteration) : ILoopEvent;

/// <summary>
/// Base class for tool-related events.
/// </summary>
public record ToolEventBase(string ToolName, Dictionary<string, object?> Parameters, int Iteration) : ILoopEvent
{
    public string Info(string emoji)
    {
        if (Parameters.Count == 0)
            return $"{emoji} {ToolName}";

        var values = string.Join(", ", Parameters.Values.Select(v => v?.ToString() ?? "null"));
        return $"{emoji} {ToolName}: {values}";
    }
}

/// <summary>
/// Indicates a tool execution has started.
/// </summary>
public sealed record ToolExecutionStartEvent(string ToolName, Dictionary<string, object?> Parameters, int Iteration)
    : ToolEventBase(ToolName, Parameters, Iteration);

/// <summary>
/// Indicates a tool was executed.
/// </summary>
public sealed record ToolExecutionEvent(
    string ToolName,
    Dictionary<string, object?> Parameters,
    string? Result,
    Exception? Error,
    TimeSpan Duration,
    int Iteration
) : ToolEventBase(ToolName, Parameters, Iteration);

/// <summary>
/// Indicates the promise phrase was found.
/// </summary>
public sealed record PromiseDetectedEvent(string Phrase, string Source, int Iteration) : ILoopEvent;

/// <summary>
/// Indicates an error occurred.
/// </summary>
public sealed record ErrorEvent(Exception Error, int Iteration, bool Recoverable) : ILoopEvent;
